service/worker/archiver/archiver.go

```diff
@@ -0,0 +1,166 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package archiver
+
+import (
+	"time"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/metrics"
+	"go.uber.org/cadence"
+	"go.uber.org/cadence/workflow"
+)
+
+type (
+	// Archiver is used to process archival requests
+	Archiver interface {
+		Start()
+		Finished() []uint64
+	}
+
+	archiver struct {
+		ctx           workflow.Context
+		logger        bark.Logger
+		metricsClient metrics.Client
+		concurrency   int
+		requestCh     workflow.Channel
+		resultCh      workflow.Channel
+	}
+)
+
+// NewArchiver returns a new Archiver
+func NewArchiver(
+	ctx workflow.Context,
+	logger bark.Logger,
+	metricsClient metrics.Client,
+	concurrency int,
+	requestCh workflow.Channel,
+) Archiver {
+	return &archiver{
+		ctx:           ctx,
+		logger:        logger,
+		metricsClient: metricsClient,
+		concurrency:   concurrency,
+		requestCh:     requestCh,
+		resultCh:      workflow.NewBufferedChannel(ctx, concurrency),
+	}
+}
+
+// Start spawns concurrency count of coroutine to handle archivals (does not block).
+func (a *archiver) Start() {
+	a.metricsClient.IncCounter(metrics.ArchiverScope, metrics.ArchiverStartedCount)
+	for i := 0; i < a.concurrency; i++ {
+		workflow.Go(a.ctx, func(ctx workflow.Context) {
+			a.metricsClient.IncCounter(metrics.ArchiverScope, metrics.ArchiverCoroutineStartedCount)
+			var handledHashes []uint64
+			for {
+				var request ArchiveRequest
+				if more := a.requestCh.Receive(ctx, &request); !more {
+					break
+				}
+				handleRequest(ctx, a.logger, a.metricsClient, request)
+				handledHashes = append(handledHashes, hashArchiveRequest(request))
+			}
+			a.resultCh.Send(a.ctx, handledHashes)
+			a.metricsClient.IncCounter(metrics.ArchiverScope, metrics.ArchiverCoroutineStoppedCount)
+		})
+	}
+}
+
+// Finished will block until all work has been finished.
+// Returns hashes of requests handled.
+func (a *archiver) Finished() []uint64 {
+	var handledHashes []uint64
+	for i := 0; i < a.concurrency; i++ {
+		var subResult []uint64
+		a.resultCh.Receive(a.ctx, &subResult)
+		handledHashes = append(handledHashes, subResult...)
+	}
+	a.resultCh.Close()
+	a.metricsClient.IncCounter(metrics.ArchiverScope, metrics.ArchiverStoppedCount)
+	return handledHashes
+}
+
+func handleRequest(ctx workflow.Context, logger bark.Logger, metricsClient metrics.Client, request ArchiveRequest) {
```

Comment:
I agree it does but it also makes it very easy to write a bug. The context given to this method must be the context from the callback in which this function is invoked rather than the context that is a member field on this object. I decided to not make it a member function in order to protect against this bug.